File: 198-house-robber/house-robber.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        # memoize each best result
        # to calculate the current house look at the previous adjacent house and the previous before
        # and add it to the value of the current house

        if len(nums) == 1:
            return nums[0]
        if len(nums) == 2:
            return max(nums)
        memo = [0] * len(nums)
        memo[0] = nums[0]
        memo[1] = nums[1]
        memo[2] = memo[0] + nums[2]

        for current in range(3, len(nums)):
            memo[current] = max(memo[current - 2], memo[current - 3]) + nums[current]

        return max(memo)

File: 198-house-robber/house-robber.java
class Solution {
    public int rob(int[] nums) {
        int[] memo = new int[nums.length];
        switch (nums.length) {
            case 0:
                return 0;
            case 1:
                return nums[0];
            case 2:
                return Math.max(nums[0], nums[1]);
            default:
                break;
        }
        memo[0] = nums[0];
        memo[1] = Math.max(nums[0], nums[1]);
        for(int x = 2; x < nums.length; x++) {
            memo[x] = Math.max(memo[x -1 ], memo[x - 2] + nums[x]);
        }

        return memo[nums.length - 1];
    }
}